pkg/actions/lua/lakefs/catalogexport/hive.lua

```diff
@@ -0,0 +1,100 @@
+local regexp = require("regexp")
+local pathlib = require("path")
+local common = require("lakefs/catalogexport/common")
+
+-- extract partition prefix from full path
+function extract_partitions_path(partition_cols, path)
+    -- list of columns to pattern {a,b,c} -> a=*/b=*/c=*/
+    local partition_pattern = table.concat(partition_cols, "=[^/]*/") .. "=[^/]*/"
+    local re = regexp.compile(partition_pattern)
+    local match = re.find(path, partition_pattern)
+    if match == "" then
+        return nil
+    end
+    return match
+end
+
+-- Hive format partition iterator each result set is a collection of files under the same partition
+function lakefs_hive_partition_pager(client, repo_id, commit_id, base_path, page_size, partition_cols)
+    -- local has_more = true
+    local prefix = base_path
+    local target_partition = ""
+    local pager = common.lakefs_object_pager(client, repo_id, commit_id, "", prefix, page_size, "")
+    local page = pager()
+    return function()
+        if page == nil then
+            return nil
+        end
+        local partition_entries = {}
+        repeat
+            if #page == 0 then
+                page = pager()
+                if page == nil then -- no more records
+                    return target_partition, partition_entries
+                end
+            end
+            -- repeat
+            local entry = page[1]
+            if not pathlib.is_hidden(entry.path) then
+                local partition_key = extract_partitions_path(partition_cols, entry.path)
+                -- first time: if not set, assign current object partition as the target_partition key
+                if target_partition == "" then
+                    target_partition = partition_key
+                end
+                -- break if current entry does not belong to the target_partition
+                if partition_key ~= target_partition then
+                    local partition_result = target_partition
+                    target_partition = partition_key
+                    return partition_result, partition_entries
+                end
+                table.insert(partition_entries, {
+                    physical_address = entry.physical_address,
+                    path = entry.path,
+                    size = entry.size_bytes,
+                    checksum = entry.checksum
+                })
+                -- remove entry only if its part of the current partition
+                table.remove(page, 1)
+            end
+        until not true -- check if has while True 
```

Comment:
Prefer `while true do ... end` it will be more readable top-down as endless loop

Reply:
👍